File: week1/process_packages/process_packages.py
# python3

from collections import namedtuple

Request = namedtuple("Request", ["arrived_at", "time_to_process"])
Response = namedtuple("Response", ["was_dropped", "started_at"])


class Buffer:
    def __init__(self, size):
        self.size = size
        self.finish_time = []

    def process(self, request):
        # write your code here
        if len(self.finish_time) == 0:
            self.finish_time.append(request.arrived_at + request.time_to_process)
            #print("Buffer (queue of finish_time_) is empty, push new process in queue and process it intermediately")
            return Response(False, request.arrived_at)
        else:
            if request.arrived_at >= self.finish_time[0]:
                finish_time_back = self.finish_time[len(self.finish_time) - 1]
                if request.arrived_at <= finish_time_back:
                    #print("arrival_time of request is latter than finish_time of first process in queue. But early than last process in queue")
                    self.finish_time.append(finish_time_back + request.time_to_process)
                    self.finish_time.pop(0)
                    return Response(False, finish_time_back)
                else:
                    #print("arrival_time of request is latter than finish_time of first process in queue. and latter than last process in queue")
                    self.finish_time.append(request.arrived_at + request.time_to_process)
                    self.finish_time.pop(0)
                    return Response(False, request.arrived_at)
            else:
                if len(self.finish_time) < self.size:
                    finish_time_back = self.finish_time[len(self.finish_time) - 1]
                    #print("arrival_time of request is early than finish_time of first process in queue. But there is still space in buffer, message will not be dropped")
                    self.finish_time.append(finish_time_back + request.time_to_process)
                    return Response(False, finish_time_back)
                else:
                    #print("arrival_time of request is early than finish_time of first process in queue. And there is no space in buffer, message will be dropped")
                    return Response(True, -1)

def process_requests(requests, buffer):
    responses = []
    for request in requests:
        responses.append(buffer.process(request))
    return responses


def main():
    buffer_size, n_requests = map(int, input().split())
    requests = []
    for _ in range(n_requests):
        arrived_at, time_to_process = map(int, input().split())
        requests.append(Request(arrived_at, time_to_process))

    buffer = Buffer(buffer_size)
    responses = process_requests(requests, buffer)

    for response in responses:
        print(response.started_at if not response.was_dropped else -1)


if __name__ == "__main__":
    main()

File: week1/process_packages/process_packages.cpp
#include <iostream>
#include <queue>
#include <vector>

struct Request {
    Request(int arrival_time, int process_time):
        arrival_time(arrival_time),
        process_time(process_time)
    {}

    int arrival_time;
    int process_time;
};

struct Response {
    Response(bool dropped, int start_time):
        dropped(dropped),
        start_time(start_time)
    {}

    bool dropped;
    int start_time;
};

class Buffer {
public:
    Buffer(int size):
        size_(size),
        finish_time_()
    {}

    Response Process(const Request &request) {
        // write your code here
        Response response(false, -1);
        if (finish_time_.empty() && size_ > 0){
            response.dropped = false;
            response.start_time = request.arrival_time;
            finish_time_.push(request.arrival_time + request.process_time);
            //std::cout << "Buffer (queue of finish_time_) is empty, push new process in queue and process it intermediately" << std::endl;
            return response;
        }
        else {
            if (request.arrival_time >= finish_time_.front()){
                response.dropped = false;
                if (request.arrival_time <= finish_time_.back()){
                    response.start_time = finish_time_.back();
                    finish_time_.push(finish_time_.back() + request.process_time);
                    //std::cout << "arrival_time of request is latter than finish_time of first process in queue. But early than last process in queue" << std::endl;                    
                }
                else{
                    response.start_time = request.arrival_time;
                    //std::cout << "arrival_time of request is latter than finish_time of first process in queue. and latter than last process in queue" << std::endl;
                    finish_time_.push(request.arrival_time + request.process_time);
                }
                //finish_time_.push(finish_time_.back() + request.process_time);
                finish_time_.pop();
                return response;
           }
            else {
                if (finish_time_.size() < size_){
                    response.dropped = false;
                    response.start_time = finish_time_.back();
                    finish_time_.push(finish_time_.back() + request.process_time);
                    //std::cout << "arrival_time of request is early than finish_time of first process in queue. But there is still space in buffer, message will not be dropped" << std::endl;
                    return response;
                }
                else {
                    response.dropped = true;
                    //std::cout << "arrival_time of request is early than finish_first of last process in queue. And there is no space in buffer, message will be dropped" << std::endl;
                    return response;
                }
            }
        }
    }
private:
    int size_;
    std::queue <int> finish_time_;
};

std::vector <Request> ReadRequests() {
    std::vector <Request> requests;
    int count;
    std::cin >> count;
    for (int i = 0; i < count; ++i) {
        int arrival_time, process_time;
        std::cin >> arrival_time >> process_time;
        requests.push_back(Request(arrival_time, process_time));
    }
    return requests;
}

std::vector <Response> ProcessRequests(const std::vector <Request> &requests, Buffer *buffer) {
    std::vector <Response> responses;
    for (int i = 0; i < requests.size(); ++i)
        responses.push_back(buffer->Process(requests[i]));
    return responses;
}

void PrintResponses(const std::vector <Response> &responses) {
    for (int i = 0; i < responses.size(); ++i)
        std::cout << (responses[i].dropped ? -1 : responses[i].start_time) << std::endl;
}

int main() {
    int size;
    std::cin >> size;
    std::vector <Request> requests = ReadRequests();

    Buffer buffer(size);
    std::vector <Response> responses = ProcessRequests(requests, &buffer);

    PrintResponses(responses);
    return 0;
}
